ship template function docs (#822)
replicatedhq/help-center: content/docs/ship/template-functions/template-types.md

---
date: "2018-05-02T01:19:20Z"
title: "Template Function Types"
description: "Types of template functions available in Ship"
weight: "45002"
categories: [ "Ship Template Functions" ]
index: ["docs/products", "docs"]
icon: "replicatedShip"
gradient: "console"
---

{{< linked_headline "Replicated Ship Template Functions" >}}

Template functions are the key to providing dynamic contents or behavior within a Ship application.

> They can be used to generated the contents of rendered files, skip unneeded steps, and provide more relevant messaging to users, along with a multitude of other things.

Reviewer: ```suggestion
They can be used to generates the contents of rendered files, skip unneeded steps, and provide more relevant messaging to users, along with a multitude of other things.
```

Ship template functions extend those available within the [text/template](https://golang.org/pkg/text/template/) package of Go.

> All Ship template functions must be escaped with `{{repl`, rather than the standard `{{`.

Reviewer: ```suggestion
All Ship template functions must begin with `{{repl`, rather than the standard `{{`. See [Static Functions](#static-functions) for some examples.
```

{{< linked_headline "Template Function List" >}}

The many template functions available within Ship can be thought of as belonging to several categories.

> # Sprig

Reviewer: I believe our contributing.md [forbids H1 tags](https://github.com/replicatedhq/help-center/blob/master/CONTRIBUTING.md#avoid-h1-tags-on-a-page)

Author: ```suggestion
## Sprig
```

Author: Actually @dexhorthy should I use `{{< linked_headline "Sprig" >}}` here (and for the other headings?)

Many of the utility functions provided come from Sprig, a third-party library of Go template functions. 
The Sprig documentation can be found [here](https://masterminds.github.io/sprig/).

> # Static functions

Author: ```suggestion
## Static functions
```

```go
func ParseFloat(str string) float64
```
ParseFloat returns the float value represented by the string.
```yaml
'{{repl ConfigOption "str_value" | ParseFloat }}'
```

{{< template_function name="ParseInt" >}}
```go
func ParseInt(str string, args ...int) int64
```
ParseInt returns the integer value represented by the string with optional base (default 10).
```yaml
'{{repl ConfigOption "str_value" | ParseInt }}'
```

{{< template_function name="ParseUint" >}}
```go
func ParseUint(str string, args ...int) uint64
```
ParseUint returns the unsigned integer value represented by the string with optional base (default 10).
```yaml
'{{repl ConfigOption "str_value" | ParseUint }}'
```

{{< template_function name="HumanSize" >}}
```go
func HumanSize(size interface{}) string
```
HumanSize returns a human-readable approximation of a size in bytes capped at 4 valid numbers (eg. "2.746 MB", "796 KB"). The size must be a integer or floating point number.
```yaml
'{{repl ConfigOption "min_size_bytes" | HumanSize }}'
```

{{< template_function name="KubeSeal" >}}
```go
func KubeSeal(certData string, namespace string, name string, value string) string
```

> # Installation Context Functions

Author: ```suggestion
## Installation Context Functions
```

Reviewer: is this suggestion missing a leading `{` ?

```go
func ShipCustomerRelease() string
```
ShipCustomerRelease returns a copy of the release metadata struct, marshalled as yaml.
```yaml
'{{repl ShipCustomerRelease }}'
```

{{< template_function name="EntitlementValue" >}}
```go
func EntitlementValue(name string) string
```
EntitlementValue returns the value of the entitlement with the provided name.
```yaml
'{{repl EntitlementValue "numSeats" }}'
```

{{< template_function name="LicenseFieldValue" >}}
```go
func LicenseFieldValue(name string) string
```
LicenseFieldValue returns the value of the entitlement with the provided name. Is an alias for EntitlementValue.
```yaml
'{{repl LicenseFieldValue "numSeats" }}'
```

{{< template_function name="Installation" >}}
```go
func Installation(field string) string
```
Possible Options:

| Key | Type |
| --- | ---- |
| state_file_path | string |
| customer_id | string |
| semver | string |
| channel_name | string |
| channel_id | string |
| release_id | string |
| installation_id | string |
| release_notes | []string |
| app_slug | []string |
| license_id | string |

Installation returns the value of the property with the provided key.
```yaml
'{{repl Installation "license_id" }}'
```

> # Config Context Functions

Author: ```suggestion
## Config Context Functions
```

Reviewer: ditto `{`

{{< template_function name="ConfigOption" >}}
```go
func ConfigOption(optionName string) string
```
Returns the value of the config option as a string.
The config screen and associated options are described [here](/docs/ship/config/overview).
```yaml
'{{repl ConfigOption "hostname" }}'
```

{{< template_function name="ConfigOptionData" >}}

```go
func ConfigOptionData(fileName string) string
```
Returns the base64 decoded value of a config option.
```yaml
'{{repl ConfigOptionData "ssl_key"}}'
```

{{< template_function name="ConfigOptionEquals" >}}
```go
func ConfigOptionEquals(optionName string, expectedValue string) bool
```
Returns true if the configuration option value is equal to the supplied value.
```yaml
'{{repl ConfigOptionEquals "http_enabled" "1" }}'
```

{{< template_function name="ConfigOptionNotEquals" >}}
```go
func ConfigOptionNotEquals(optionName string, expectedValue string) bool
```
Returns true if the configuration option value is not equal to the supplied value.
```yaml
'{{repl ConfigOptionNotEquals "http_enabled" "1" }}'
```

> # Ship Context Functions

Author: ```suggestion
## Ship Context Functions
```

```

{{< template_function name="GetCaCert" >}}

```go
func GetCaCert(caName string) string
```
Returns a cert for a previously generated certificate authority.
Must be called after `GetCaKey` has been called for the desired CA name.
```yaml
'{{repl GetCaCert "my_certificate_authority"}}'
```

{{< template_function name="GetKey" >}}

```go
func GetKey(certName string, caName string, hosts string, certType string) string
```
Returns a key for a cert with the desired properties.
`hosts` is a set of hosts that the certificate should be valid for, seperated by commas, such as `example1.com,subdomain.example.io`.
If a cert with this name has been requested before, the same value will be returned, even if other request parameters differ.
If a CA with the desired name has been requested before, that CA will be used to sign the generated cert for the requested hosts.
If a CA with the desired name has not been requested before, one will be generated with the same type as this cert.
The type can be RSA or ECDSA, with acceptable values being `rsa-2048`, `rsa-4096`, `rsa-8192`, `P256` or `P521`.
If no type is provided, `rsa-2048` will be used.
```yaml

> '{{repl GetKey "my_certificate_authority" "P521"}}'

Reviewer: This appears to missing the "hosts" parameter?

Author: 🤦‍♂ 
And the CA name parameter too

> '{{repl GetKey "my_certificate_authority" "P521"}}'

Reviewer: Should these use a `certName` other than `my_certificate_authority`? To distiguish CA names from Cert names?

Reviewer: yeah you got it

Author: yeah, it's a 'copy-paste-and-then-not-update' error

> '{{repl GetKey "my_cert" "my_certificate_authority" "example.com,myexampleapp.io,help.replicated.com" "P521"}}'

Reviewer: one question I would have -- can I use this to get a wildcard cert?

Author: I think so? This just uses cfssl under the hood
